File: grind75/Dec30/200.go
func numIslands(grid [][]byte) int {
    m,n := len(grid), len(grid[0])
    dirs := [][]int{[]int{1,0}, []int{-1,0}, []int{0,1}, []int{0,-1},}
    visited := make([]int, m*n)
    ret := 0   
    var dfs func(int, int)
    
    dfs = func(i,j int) {
        for _, dir := range dirs {
            ni, nj := i+dir[0], j+dir[1]
            if ni < 0 || ni >= m || nj < 0 || nj >= n || visited[ni*n+nj] == 1 || grid[ni][nj] == '0' {
                continue
            }
            visited[ni*n+nj] = 1
            dfs(ni, nj)
        }
    }

    for i := 0; i < m; i++ {
        for j := 0; j < n; j++ {
            if visited[i*n+j] == 1 || grid[i][j] == '0' {
                continue
            }
            ret += 1
            dfs(i,j)
        }
    }

    return ret
}

File: grind75/Dec29/200.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:

        # dfs
        # O(mn), O(mn)
        m = len(grid)
        n = len(grid[0])
        count = 0
        def dfs(grid, i, j):

            if i < 0 or i >= m or j < 0 or j >= n or grid[i][j] != '1':
                return
            
            grid[i][j] = '0'

            dfs(grid, i+1, j)
            dfs(grid, i-1, j)
            dfs(grid, i, j+1)
            dfs(grid, i, j-1)
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    count += 1
                    dfs(grid, i, j)
        
        return count
